File: src/NodeParser/TypeOperatorNodeParser.ts
import ts from "typescript";
import { Context, NodeParser } from "../NodeParser.js";
import { SubNodeParser } from "../SubNodeParser.js";
import { ArrayType } from "../Type/ArrayType.js";
import { BaseType } from "../Type/BaseType.js";
import { NumberType } from "../Type/NumberType.js";
import { ObjectType } from "../Type/ObjectType.js";
import { StringType } from "../Type/StringType.js";
import { UnionType } from "../Type/UnionType.js";
import { derefType } from "../Utils/derefType.js";
import { getTypeKeys } from "../Utils/typeKeys.js";

export class TypeOperatorNodeParser implements SubNodeParser {
    public constructor(protected childNodeParser: NodeParser) {}

    public supportsNode(node: ts.TypeOperatorNode): boolean {
        return node.kind === ts.SyntaxKind.TypeOperator;
    }

    public createType(node: ts.TypeOperatorNode, context: Context): BaseType {
        const type = this.childNodeParser.createType(node.type, context);
        const derefed = derefType(type);
        // Remove readonly modifier from type
        if (node.operator === ts.SyntaxKind.ReadonlyKeyword && derefed) {
            return derefed;
        }
        if (derefed instanceof ArrayType) {
            return new NumberType();
        }
        const keys = getTypeKeys(type);
        if (derefed instanceof ObjectType && derefed.getAdditionalProperties()) {
            return new UnionType([...keys, new StringType()]);
        }

        if (keys.length === 1) {
            return keys[0];
        }

        return new UnionType(keys);
    }
}

File: internal/parser/type_operator_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// TypeOperatorNodeParser parses `keyof`/`readonly`/`unique` type operator
// nodes (src/NodeParser/TypeOperatorNodeParser.ts).
type TypeOperatorNodeParser struct {
	childNodeParser NodeParser
}

func NewTypeOperatorNodeParser(childNodeParser NodeParser) *TypeOperatorNodeParser {
	return &TypeOperatorNodeParser{childNodeParser: childNodeParser}
}

func (p *TypeOperatorNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindTypeOperator
}

func (p *TypeOperatorNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	operatorNode := node.AsTypeOperatorNode()
	typ := p.childNodeParser.CreateType(operatorNode.Type, ctx, nil)
	derefed := types.DerefType(typ)

	// Remove readonly modifier from type.
	if operatorNode.Operator == ast.KindReadonlyKeyword && derefed != nil {
		return derefed
	}

	if _, isArray := derefed.(*types.ArrayType); isArray {
		return &types.NumberType{}
	}

	keys := types.GetTypeKeys(typ)
	keyTypes := make([]types.Type, len(keys))
	for i, key := range keys {
		keyTypes[i] = key
	}

	if object, isObject := derefed.(*types.ObjectType); isObject {
		if _, hasTypedAdditional := object.AdditionalProperties.(types.Type); hasTypedAdditional {
			return types.NewUnionType(append(keyTypes, &types.StringType{}))
		}
	}

	if len(keys) == 1 {
		return keys[0]
	}

	return types.NewUnionType(keyTypes)
}
